### DISCUSSION TITLE
hu values

### DISCUSSION THREAD

**Discussion by kalleick** [dewenni/ESP_Buderus_KM271#26] [Q&A] [answered]
I implemented option 2 (with original KM271) and ESP32.
 I get the setting values in the dashboard (e.g. automatic operating mode, burner on/off), but all temperature values are marked with **"hu".** 
So the communication seems to work in principle. 

**But why am I not getting any readings?** 

Does it take a long time to see values or is there a defect?

![2023-05-06 13_05_27-Buderus Logamatic](https://user-images.githubusercontent.com/95974901/236620452-1ccbabde-5532-432d-9575-de63e509d7d7.png)

**Comment by dewenni** [comment]
Hi,

that is interesting. I had a similar behavior in the past.
I am pretty sure that this is related to the printf format strings that I am using.

Are you using the latest version?

the related code is this:
<img width="985" alt="image" src="https://user-images.githubusercontent.com/46074831/236630425-4ad1158c-dbfd-4841-a438-aa37d47048d3.png">

the function is located in basic.cpp:
```c
const char* uint8ToString(uint8_t value){
  static char ret_str[4];
  snprintf(ret_str, sizeof(ret_str), "%hhu", value);
  return ret_str;
}   
```

In your setup there must be some difference regarding the data types and the formatting of the printf.
But I honestly don't know what that depends on.

I think I must find a different solution for that conversions from number to string that is more robust.

**Reply by dewenni** [reply]
@kalleick 
I did some research and I would like to ask you to try this code change:

add `#include <inttypes.h>` to the `basic.h` file

change the code of the convert functions in `basic.cpp` like below - using `PRIi8`, `PRIu8`, `PRIu16`, `PRIu64`

```c
const char* int8ToString(int8_t value){
  static char ret_str[4];
  snprintf(ret_str, sizeof(ret_str), "%" PRIi8 "", value);
  return ret_str;
}

const char* uint8ToString(uint8_t value){
  static char ret_str[4];
  snprintf(ret_str, sizeof(ret_str), "%" PRIu8 "", value);
  return ret_str;
}

const char* uint16ToString(uint16_t value){static char ret_str[10];
  snprintf(ret_str, sizeof(ret_str), "%" PRIu16 "", value);
  return ret_str;
}

const char* uint64ToString(uint64_t value){
  static char ret_str[64];
  snprintf(ret_str, sizeof(ret_str), "%" PRIu64 "", value);
  return ret_str;
}
```

please give me feedback if this works for you. If yes, I will change the code and build a new release version.

Thanks!

**Reply by kalleick** [reply]
Hello Sven,

great, thanks for the quick help, that's really great!

yes i have had the newest version. my project started today 


I have now done the following steps:
1. basics.h supplemented by the #include <inttypes.h>
2. Code changed
3. Build successful
4. OTA firmware upload from new file "firmware.bin"
5. Reboot
But no change, still "hu" where the values should be.
Maybe I have a too old Logamatic version (3.18)?

One more note: via MQTT I get these values:
 
2023-05-06_21:29:41 MQTT2_ESP_Buderus_KM271 ww_temp: 58
2023-05-06_21:29:41 MQTT2_ESP_Buderus_KM271 boiler_temp: 44
the others are also "hu", see attachment.

Best regards
Patrick
[fhem_mqtt.txt](https://github.com/dewenni/ESP_Buderus_KM271/files/11413
![2023-05-06 21_36_12-Buderus Logamatic](https://user-images.githubusercontent.com/95974901/236643459-f52d3124-523f-4818-bfa3-e24970dbf5aa.png)
225/fhem_mqtt.txt)
![2023-05-06 21_25_15-Buderus Logamatic](https://user-images.githubusercontent.com/95974901/236643430-4a51bcad-029b-419f-ad1b-a887d8c7eebb.png)

**Reply by dewenni** [reply]
a pity...

but I still think that it is related to this format specifiers and your esp board.
here has someone else reported the same problem: https://github.com/HelTecAutomation/CubeCell-Arduino/issues/117

It seems that in some cases the length specifier `hh` or `ll` is not supported.


then please try this one

```c
const char* uint8ToString(uint8_t value){
  static char ret_str[4];
  snprintf(ret_str, sizeof(ret_str), "%u", value);
  return ret_str;
}
```

or maybe it is even better to also cast the value to unsigned int

```c
const char* uint8ToString(uint8_t value){
  static char ret_str[4];
  snprintf(ret_str, sizeof(ret_str), "%u", (unsigned int)value);
  return ret_str;
}
```

can you please try and give me feedback?

**Comment by kalleick** [comment]
thank you, very much, this one worked:

const char* uint8ToString(uint8_t value){
  static char ret_str[4];
  snprintf(ret_str, sizeof(ret_str), "%u", value);
  return ret_str;
}


info: i have this board:
https://amzn.eu/d/9RxEe0u
![2023-05-06 22_44_02-Buderus Logamatic](https://user-images.githubusercontent.com/95974901/236645905-189faa8c-c499-4d67-ad54-81c30d7f3c97.png)

**Reply by dewenni** [reply]
ok, than I should change it that way.

but two more questions.

is this shown correctly?
<img width="736" alt="image" src="https://user-images.githubusercontent.com/46074831/236646003-e2433cad-9891-4084-ad9f-99469464d544.png">


and also this one?
<img width="587" alt="image" src="https://user-images.githubusercontent.com/46074831/236646020-fb2217ba-6b04-4a85-8b25-b4a056485207.png">

**Comment by kalleick** [comment]
hi Sven, thank you so much, yes, some values are declared with "hi", see my screenshots.
MQTT Port ist OK. 

![2023-05-07 08_22_42-Buderus Logamatic](https://user-images.githubusercontent.com/95974901/236661442-81990d7d-aef8-4b73-82d2-2e07c821f9a8.png)
![2023-05-07 08_22_28-Buderus Logamatic](https://user-images.githubusercontent.com/95974901/236661444-c5575c0d-1ff8-4890-adf0-4cfd5fe25906.png)
![2023-05-07 08_22_13-Buderus Logamatic](https://user-images.githubusercontent.com/95974901/236661445-dcc9c7b2-6bef-43c4-8236-20a276caebd9.png)
![2023-05-07 08_21_36-Buderus Logamatic](https://user-images.githubusercontent.com/95974901/236661446-3c2aeb77-7498-4b62-9a42-ce8a79f29e48.png)

**Reply by dewenni** [reply, marked as answer]
ok, its pretty clear that in your setup the 3-sign format specifier (hhi, hhu, llu) are not supported, even if this should be the right ones.
A format specifier %hhi result in printing hi, %hhu result in printing hu and %llu result in printing lu
Also if you use the combination with <inttypes.h> and the generic format specifier it seems that they are replaced with (hhi, hhu, llu) by the compiler that do not work.

I will change the functions to the "basic" format specifier (%u and %i) and hope that this will work for everyone else as well.
For the 64bit value, I will implement a manual way.
